### PR TITLE
rdm: provide test writing recommendation

### PR DESCRIPTION
**Pull request by miri64**
Memo to explain what the difference is between `unittests`, `tests`, and `examples` and what to write to test a certain feature.

### CODE REVIEW

File: doc/memos/rdm4.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  testing of provided changes.
+
+## 1. Introduction
+At the moment of writing there are two strategies for testing: providing a
+unittest suite in [`tests/unittests/`][unittests] or a stand-alone test
+application in `tests/` directly. This document provides some recommendations
+on where to put a test and what the minimum requirements for those are.
+
+## 2. Unittests
+Unittests in RIOT are provided using a costumized fork of the [embUnit]
```
Comment (kaspar030) [inline doc/memos/rdm4.md]: customized

File: doc/memos/rdm4.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+framework. While it is possible to provide a stand-alone application using this
+framework it is recommended to add or extend a test suite in the [unittests]
+application (see [unittests README] for how to do that).
+
+The advantage of that is that this application is at the time of writing this
```
Comment (kaspar030) [inline doc/memos/rdm4.md]: some words too much. I'd drop the description of the current state, as that will hopefully change soon. No need to write things that will be obsolete in the short term.

File: doc/memos/rdm4.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+comprehensive [pexpect] script that may utilize the [testrunner] library.
+
+## 3. Examples
+In the RIOT source tree there is also an `examples/` directory. This contains
+applications that exemplify features of RIOT. These applications are supposed to
+be well documented and should show-case features ideally very concentrated (i.e.
+as few features as possible - ideally one - are show-cased in the application).
+If an example is extensive enough to provide an acceptable level of coverage,
+they can be provided alternatively to a stand-alone test application, but it
+shall not be main motivation to provide an example or for a show-case in an
```
Comment (kaspar030) [inline doc/memos/rdm4.md]: wording is a little off

**Comment by miri64** [comment]
Comments addressed.

**Comment by emmanuelsearch** [comment]
addressed all comments.
(will have to fix commit history later)

File: doc/memos/rdm4.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+## Status
+This document is a product of the community of RIOT maintainers, and aims to 
+represent the consensus within this community.
+The content of this document is licensed with a Creative Commons CC-BY-SA 
+license.
+
+## Terminology
+This memo uses the following terminology and acronyms:
+
+- embUnit: a unittest framework for C that is used by RIOT to write unittests.
```
Comment (emmanuelsearch) [inline doc/memos/rdm4.md]: "the unit test..."?
https://en.wikipedia.org/wiki/Unit_testing

File: doc/memos/rdm4.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+- pexpect: a python library to interact with the stdio of command-line
+  application.
+- CI system: CI = Continuous Intergration, platform for automated building and
+  testing of provided changes.
+
+## 1. Introduction
+At the moment of writing there are two strategies for testing: providing a
```
Comment (emmanuelsearch) [inline doc/memos/rdm4.md]: I would remove "at the moment of writing".

File: doc/memos/rdm4.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+applicable components and one or multiple stand-alone tests for the others.
+
+A stand-alone test application should be build-able for all supported platforms
+and, if they fit, also be usable on these platforms. To simplify and streamline
+testing efforts for other developers it is also recommended to provide a
+comprehensive [pexpect] script that may utilize the [testrunner] library.
+
+## 3. Examples
+In the RIOT source tree there is also an `examples/` directory. This contains
+applications that exemplify features of RIOT. These applications are supposed to
```
Comment (emmanuelsearch) [inline doc/memos/rdm4.md]: "showcase" instead of "exemplify"?

File: doc/memos/rdm4.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  application.
+- CI system: CI = Continuous Intergration, platform for automated building and
+  testing of provided changes.
+
+## 1. Introduction
+At the moment of writing there are two strategies for testing: providing a
+unittest suite in [`tests/unittests/`][unittests] or a stand-alone test
+application in `tests/` directly. This document provides some recommendations
+on where to put a test and what the minimum requirements for those are.
+
```
Comment (emmanuelsearch) [inline doc/memos/rdm4.md]: would it make sense to indicate at what point (level of granularity) one should consider providing tests?
Something like "if thou providest a new feature, thou shalt provide a test".
Then, if providing tests are to be provided, then the question is indeed unit test vs test in example.

**Comment by miri64** [comment]
Addressed comments

**Comment by PeterKietzmann** [comment]
@miri64 @emmanuelsearch I will announce feature freeze in about 1-2 hours. You think you can get this done by then?

**Comment by miri64** [comment]
Closed as memo, since original RDM PR was closed, too.

**Pull request closed without merging**